File: ios/CommuteOptimizer/CommuteOptimizer/Utilities/MtaColors.swift
import SwiftUI

struct MtaColors {
    // Official MTA line colors
    static let lineColors: [String: Color] = [
        // Red lines (1, 2, 3)
        "1": Color(hex: "#EE352E"),
        "2": Color(hex: "#EE352E"),
        "3": Color(hex: "#EE352E"),

        // Green lines (4, 5, 6)
        "4": Color(hex: "#00933C"),
        "5": Color(hex: "#00933C"),
        "6": Color(hex: "#00933C"),

        // Purple line (7)
        "7": Color(hex: "#B933AD"),

        // Blue lines (A, C, E)
        "A": Color(hex: "#0039A6"),
        "C": Color(hex: "#0039A6"),
        "E": Color(hex: "#0039A6"),

        // Orange lines (B, D, F, M)
        "B": Color(hex: "#FF6319"),
        "D": Color(hex: "#FF6319"),
        "F": Color(hex: "#FF6319"),
        "M": Color(hex: "#FF6319"),

        // Lime green line (G)
        "G": Color(hex: "#6CBE45"),

        // Brown lines (J, Z)
        "J": Color(hex: "#996633"),
        "Z": Color(hex: "#996633"),

        // Gray line (L)
        "L": Color(hex: "#A7A9AC"),

        // Yellow lines (N, Q, R, W)
        "N": Color(hex: "#FCCC0A"),
        "Q": Color(hex: "#FCCC0A"),
        "R": Color(hex: "#FCCC0A"),
        "W": Color(hex: "#FCCC0A"),

        // Shuttle (S)
        "S": Color(hex: "#808183")
    ]

    static func color(for line: String) -> Color {
        // Normalize express variants (6X -> 6, 7X -> 7, FX -> F)
        let normalizedLine = cleanExpressLine(line)
        return lineColors[normalizedLine.uppercased()] ?? Color(hex: "#808183")
    }

    static func textColor(for line: String) -> Color {
        // Yellow lines need dark text
        let normalizedLine = cleanExpressLine(line)
        switch normalizedLine.uppercased() {
        case "N", "Q", "R", "W": return .black
        default: return .white
        }
    }

    /// Normalize express train variants to base line
    static func cleanExpressLine(_ line: String) -> String {
        // Express variants: 6X->6, 7X->7, FX->F, etc.
        if line.count == 2 && line.hasSuffix("X") {
            return String(line.dropLast())
        }
        return line
    }

    static func weatherEmoji(_ conditions: String, precipType: PrecipitationType) -> String {
        switch precipType {
        case .rain: return "rain"
        case .snow: return "snow"
        case .mix: return "cloud.sleet"
        case .none:
            if conditions.lowercased().contains("cloud") { return "cloud" }
            if conditions.lowercased().contains("sun") || conditions.lowercased().contains("clear") { return "sun.max" }
            return "questionmark"
        }
    }

    // Rank badge colors
    static let goldBadge = Color(hex: "#FFD700")
    static let silverBadge = Color(hex: "#C0C0C0")
    static let bronzeBadge = Color(hex: "#CD7F32")

    static func rankColor(for rank: Int) -> Color {
        switch rank {
        case 1: return goldBadge
        case 2: return silverBadge
        case 3: return bronzeBadge
        default: return Color.gray
        }
    }

    /// Badge color for arrivals - green for <=2 min, line color otherwise
    static func arrivalBadgeColor(minutesAway: Int, line: String) -> Color {
        minutesAway <= 2 ? .green : color(for: line)
    }
}

extension Color {
    init(hex: String) {
        let hex = hex.trimmingCharacters(in: CharacterSet.alphanumerics.inverted)
        var int: UInt64 = 0
        Scanner(string: hex).scanHexInt64(&int)
        let r, g, b: Double
        r = Double((int >> 16) & 0xFF) / 255.0
        g = Double((int >> 8) & 0xFF) / 255.0
        b = Double(int & 0xFF) / 255.0
        self.init(red: r, green: g, blue: b)
    }
}

File: android/app/src/main/java/com/commuteoptimizer/widget/util/MtaColors.kt
package com.commuteoptimizer.widget.util

import android.graphics.Color

object MtaColors {

    // Official MTA line colors
    private val lineColors = mapOf(
        // Red lines (1, 2, 3)
        "1" to Color.parseColor("#EE352E"),
        "2" to Color.parseColor("#EE352E"),
        "3" to Color.parseColor("#EE352E"),

        // Green lines (4, 5, 6)
        "4" to Color.parseColor("#00933C"),
        "5" to Color.parseColor("#00933C"),
        "6" to Color.parseColor("#00933C"),

        // Purple line (7)
        "7" to Color.parseColor("#B933AD"),

        // Blue lines (A, C, E)
        "A" to Color.parseColor("#0039A6"),
        "C" to Color.parseColor("#0039A6"),
        "E" to Color.parseColor("#0039A6"),

        // Orange lines (B, D, F, M)
        "B" to Color.parseColor("#FF6319"),
        "D" to Color.parseColor("#FF6319"),
        "F" to Color.parseColor("#FF6319"),
        "M" to Color.parseColor("#FF6319"),

        // Lime green line (G)
        "G" to Color.parseColor("#6CBE45"),

        // Brown lines (J, Z)
        "J" to Color.parseColor("#996633"),
        "Z" to Color.parseColor("#996633"),

        // Gray line (L)
        "L" to Color.parseColor("#A7A9AC"),

        // Yellow lines (N, Q, R, W)
        "N" to Color.parseColor("#FCCC0A"),
        "Q" to Color.parseColor("#FCCC0A"),
        "R" to Color.parseColor("#FCCC0A"),
        "W" to Color.parseColor("#FCCC0A"),

        // Shuttle (S)
        "S" to Color.parseColor("#808183")
    )

    private val defaultColor = Color.parseColor("#808183")

    fun getLineColor(line: String): Int {
        return lineColors[line.uppercase()] ?: defaultColor
    }

    fun getTextColorForLine(line: String): Int {
        // Yellow lines need dark text
        return when (line.uppercase()) {
            "N", "Q", "R", "W" -> Color.BLACK
            else -> Color.WHITE
        }
    }

    // Get weather emoji
    fun getWeatherEmoji(conditions: String, precipType: String): String {
        return when {
            precipType == "rain" -> "\uD83C\uDF27" // rain cloud
            precipType == "snow" -> "\u2744" // snowflake
            precipType == "mix" -> "\uD83C\uDF28" // cloud with snow
            conditions.contains("cloud", ignoreCase = true) -> "\u2601" // cloud
            conditions.contains("sun", ignoreCase = true) ||
                conditions.contains("clear", ignoreCase = true) -> "\u2600" // sun
            else -> ""
        }
    }

    // Get mode icon resource name
    fun getModeIcon(mode: String): String {
        return when (mode.lowercase()) {
            "bike" -> "ic_bike"
            "walk" -> "ic_walk"
            "subway" -> "ic_subway"
            else -> "ic_subway"
        }
    }

    // Clean express line suffix (e.g., "6X" -> "6")
    fun cleanExpressLine(line: String): String {
        return if (line.length == 2 && line.endsWith("X")) {
            line.dropLast(1)
        } else {
            line
        }
    }
}
